File: bridge/active-playback-session-tracker.js
/**
 * Active Playback Session Tracker
 *
 * Tracks per-client active playback sessions for the bridge/proxy.
 * Maps browser connections to opaque session identifiers (when known)
 * and provides lifecycle hooks for connect, media open/close, and disconnect.
 *
 * Security:
 * - Never stores or exposes internal SageTV sessionKey
 * - Never stores clientName:mediaFileId:openGeneration
 * - Never enumerates all server sessions
 * - Only tracks the bridge-assigned opaque connectionId → sessionId mapping
 *
 * Phase 1: The bridge cannot learn sessionId from the SageTV binary protocol
 * (it's a pure byte relay). The tracker stores the lifecycle and will return
 * the mapping once a mechanism is wired (e.g. the PWA client sends its session
 * identity via a text control message, or the server provides it via an HTTP
 * callback).
 */

const STALE_TIMEOUT_MS = 120_000; // 2 minutes without activity → mark stale
const STALE_CHECK_INTERVAL_MS = 30_000;

/**
 * @typedef {Object} TrackedSession
 * @property {string} connectionId - unique bridge connection ID
 * @property {string|null} sessionId - opaque session ID (null until learned)
 * @property {boolean} playbackActive - whether media playback is in progress
 * @property {number} lastActivityAt - epoch ms of last activity
 * @property {string} state - 'connected'|'playing'|'stale'|'disconnected'
 */

export class ActivePlaybackSessionTracker {
  constructor(options = {}) {
    /** @type {Map<string, TrackedSession>} */
    this._sessions = new Map();
    this._staleTimeoutMs = options.staleTimeoutMs || STALE_TIMEOUT_MS;
    this._staleCheckInterval = null;
    this._idCounter = 0;
  }

  /**
   * Start the stale-session reaper. Call once on bridge startup.
   */
  start() {
    if (this._staleCheckInterval) return;
    this._staleCheckInterval = setInterval(() => this._reapStale(), STALE_CHECK_INTERVAL_MS);
    if (this._staleCheckInterval.unref) this._staleCheckInterval.unref();
  }

  /**
   * Stop the reaper. Call on bridge shutdown.
   */
  stop() {
    if (this._staleCheckInterval) {
      clearInterval(this._staleCheckInterval);
      this._staleCheckInterval = null;
    }
    this._sessions.clear();
  }

  /**
   * Register a new browser connection. Returns a unique connectionId.
   * @param {object} [meta] - optional metadata (channel, host, etc.)
   * @returns {string} connectionId
   */
  onConnect(meta = {}) {
    const connectionId = `conn-${++this._idCounter}-${Date.now().toString(36)}`;
    this._sessions.set(connectionId, {
      connectionId,
      sessionId: null,
      playbackActive: false,
      lastActivityAt: Date.now(),
      state: 'connected',
      meta,
    });
    return connectionId;
  }

  /**
   * Set the opaque sessionId for a connection.
   * Called when sessionId is learned (future: from control message or server callback).
   *
   * TODO: Wire this call. The bridge currently cannot learn sessionId because:
   *   1. The binary protocol bytes are opaque (encrypted handshake, no inspection)
   *   2. No query param currently carries session identity from the PWA client
   *   3. No server-side callback mechanism exists yet
   *
   * Possible future sources:
   *   - PWA client sends {"type":"session_identity","sessionId":"..."} text frame
   *   - Server calls bridge HTTP callback on session start
   *   - Bridge inspects the MAC address from the WS query params (if added)
   *
   * @param {string} connectionId
   * @param {string} sessionId - opaque, never the internal sessionKey
   */
  setSessionId(connectionId, sessionId) {
    const entry = this._sessions.get(connectionId);
    if (!entry) return;
    entry.sessionId = sessionId;
    entry.lastActivityAt = Date.now();
  }

  /**
   * Set the clientName (MAC address) for a connection.
   * Extracted from the first binary handshake frame.
   * This is the identity the SageTV server uses to identify this MiniClient.
   *
   * @param {string} connectionId
   * @param {string} clientName - MAC address in XX:XX:XX:XX:XX:XX format
   */
  setClientName(connectionId, clientName) {
    const entry = this._sessions.get(connectionId);
    if (!entry) return;
    entry.clientName = clientName;
    entry.lastActivityAt = Date.now();
  }

  /**
   * Get the clientName for the most recently active connection.
   * Returns null if no connection has a known clientName.
   * @returns {string|null}
   */
  getActiveClientName() {
    let best = null;
    for (const entry of this._sessions.values()) {
      if (entry.state === 'disconnected' || entry.state === 'stale') continue;
      if (!entry.clientName) continue;
      if (!best || entry.lastActivityAt > best.lastActivityAt) {
        best = entry;
      }
    }
    return best ? best.clientName : null;
  }

  /**
   * Mark that playback has started for a connection.
   * @param {string} connectionId
   */
  onPlaybackStart(connectionId) {
    const entry = this._sessions.get(connectionId);
    if (!entry) return;
    entry.playbackActive = true;
    entry.state = 'playing';
    entry.lastActivityAt = Date.now();
  }

  /**
   * Mark that playback has stopped for a connection.
   * @param {string} connectionId
   */
  onPlaybackStop(connectionId) {
    const entry = this._sessions.get(connectionId);
    if (!entry) return;
    entry.playbackActive = false;
    entry.state = 'connected';
    entry.lastActivityAt = Date.now();
  }

  /**
   * Record activity (keeps the session from going stale).
   * @param {string} connectionId
   */
  onActivity(connectionId) {
    const entry = this._sessions.get(connectionId);
    if (!entry) return;
    if (entry.state === 'stale') {
      entry.state = entry.playbackActive ? 'playing' : 'connected';
    }
    entry.lastActivityAt = Date.now();
  }

  /**
   * Handle browser disconnect or SageTV socket close/error.
   * Clears mapping. Idempotent.
   * @param {string} connectionId
   */
  onDisconnect(connectionId) {
    const entry = this._sessions.get(connectionId);
    if (!entry) return; // idempotent
    entry.state = 'disconnected';
    entry.playbackActive = false;
    entry.sessionId = null;
    entry.clientName = null;
    this._sessions.delete(connectionId);
  }

  /**
   * Get the current active session with known sessionId.
   * Returns the most recently active playing session, or null.
   *
   * @returns {{ connectionId: string, sessionId: string, state: string }|null}
   */
  getActiveSession() {
    let best = null;
    for (const entry of this._sessions.values()) {
      if (entry.state === 'disconnected' || entry.state === 'stale') continue;
      if (!entry.sessionId) continue;
      if (!best || entry.lastActivityAt > best.lastActivityAt) {
        best = entry;
      }
    }
    if (!best) return null;
    return {
      connectionId: best.connectionId,
      sessionId: best.sessionId,
      state: best.state,
    };
  }

  /**
   * Get the unavailable reason if no active session is available.
   * @returns {string} reason code
   */
  getUnavailableReason() {
    if (this._sessions.size === 0) return 'no_active_session';
    for (const entry of this._sessions.values()) {
      if (entry.state !== 'disconnected') {
        if (!entry.clientName) return 'client_name_unknown';
        return 'session_id_unknown';
      }
    }
    return 'no_active_session';
  }

  /**
   * Get count of tracked connections (for diagnostics).
   * @returns {number}
   */
  get size() {
    return this._sessions.size;
  }

  /** @private */
  _reapStale() {
    const now = Date.now();
    for (const [id, entry] of this._sessions) {
      if (entry.state === 'disconnected') {
        this._sessions.delete(id);
        continue;
      }
      if (now - entry.lastActivityAt > this._staleTimeoutMs) {
        if (entry.state !== 'stale') {
          console.log(`[SessionTracker] Connection ${id} went stale (${Math.round((now - entry.lastActivityAt) / 1000)}s idle)`);
          entry.state = 'stale';
        }
      }
    }
  }
}

File: bridge-java/src/main/java/sagex/miniclient/pwa/ngcontext/ActivePlaybackSessionTracker.java
package sagex.miniclient.pwa.ngcontext;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Tracks active playback sessions for the bridge/proxy.
 * <p>
 * Maps browser WebSocket connections to opaque session identifiers (when known)
 * and provides lifecycle hooks for connect, media open/close, and disconnect.
 * <p>
 * <b>Security:</b>
 * <ul>
 *   <li>Never stores or exposes internal SageTV sessionKey</li>
 *   <li>Never stores clientName:mediaFileId:openGeneration</li>
 *   <li>Never enumerates all server sessions</li>
 *   <li>Only tracks bridge-assigned opaque connectionId → sessionId mapping</li>
 * </ul>
 * <p>
 * <b>Thread safety:</b> All public methods are thread-safe (ConcurrentHashMap + atomics).
 */
public class ActivePlaybackSessionTracker {
    private static final Logger log = LoggerFactory.getLogger(ActivePlaybackSessionTracker.class);
    private static final long DEFAULT_STALE_TIMEOUT_MS = 120_000; // 2 minutes
    private static final long STALE_CHECK_INTERVAL_MS = 30_000;

    private final ConcurrentHashMap<String, TrackedSession> sessions = new ConcurrentHashMap<>();
    private final AtomicLong idCounter = new AtomicLong(0);
    private final long staleTimeoutMs;
    private ScheduledExecutorService reaper;

    public ActivePlaybackSessionTracker() {
        this(DEFAULT_STALE_TIMEOUT_MS);
    }

    public ActivePlaybackSessionTracker(long staleTimeoutMs) {
        this.staleTimeoutMs = staleTimeoutMs;
    }

    /**
     * Start the stale-session reaper. Call once on bridge startup.
     */
    public void start() {
        if (reaper != null) return;
        reaper = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "ng-session-reaper");
            t.setDaemon(true);
            return t;
        });
        reaper.scheduleAtFixedRate(this::reapStale, STALE_CHECK_INTERVAL_MS, STALE_CHECK_INTERVAL_MS, TimeUnit.MILLISECONDS);
    }

    /**
     * Stop the reaper and clear all sessions. Call on bridge shutdown.
     */
    public void stop() {
        if (reaper != null) {
            reaper.shutdownNow();
            reaper = null;
        }
        sessions.clear();
    }

    /**
     * Register a new browser connection. Returns a unique connectionId.
     *
     * @param channel the WebSocket channel path (e.g. "/gfx", "/media")
     * @return unique connectionId
     */
    public String onConnect(String channel) {
        String connectionId = "conn-" + idCounter.incrementAndGet() + "-" + Long.toString(System.currentTimeMillis(), 36);
        TrackedSession session = new TrackedSession(connectionId, channel);
        sessions.put(connectionId, session);
        log.debug("[SessionTracker] Registered connection {} on {}", connectionId, channel);
        return connectionId;
    }

    /**
     * Set the opaque sessionId for a connection.
     * <p>
     * TODO: Wire this call. The bridge currently cannot learn sessionId because:
     * <ol>
     *   <li>The binary protocol bytes are opaque (encrypted handshake, no inspection)</li>
     *   <li>No query param currently carries session identity from the PWA client</li>
     *   <li>No server-side callback mechanism exists yet</li>
     * </ol>
     * <p>
     * Possible future sources:
     * <ul>
     *   <li>PWA client sends {"type":"session_identity","sessionId":"..."} text frame</li>
     *   <li>Server calls bridge HTTP callback on session start</li>
     *   <li>Bridge reads MAC from WS query params (if PWA adds the param)</li>
     * </ul>
     *
     * @param connectionId the connection identifier
     * @param sessionId    opaque session ID (NEVER the internal sessionKey)
     */
    public void setSessionId(String connectionId, String sessionId) {
        TrackedSession s = sessions.get(connectionId);
        if (s == null) return;
        s.sessionId = sessionId;
        s.lastActivityAt = System.currentTimeMillis();
    }

    /** Mark playback started for a connection. */
    public void onPlaybackStart(String connectionId) {
        TrackedSession s = sessions.get(connectionId);
        if (s == null) return;
        s.playbackActive = true;
        s.state = SessionState.PLAYING;
        s.lastActivityAt = System.currentTimeMillis();
    }

    /** Mark playback stopped for a connection. */
    public void onPlaybackStop(String connectionId) {
        TrackedSession s = sessions.get(connectionId);
        if (s == null) return;
        s.playbackActive = false;
        s.state = SessionState.CONNECTED;
        s.lastActivityAt = System.currentTimeMillis();
    }

    /** Record activity (prevents stale timeout). */
    public void onActivity(String connectionId) {
        TrackedSession s = sessions.get(connectionId);
        if (s == null) return;
        if (s.state == SessionState.STALE) {
            s.state = s.playbackActive ? SessionState.PLAYING : SessionState.CONNECTED;
        }
        s.lastActivityAt = System.currentTimeMillis();
    }

    /**
     * Handle browser disconnect or SageTV socket close/error.
     * Clears mapping. Idempotent — safe to call multiple times.
     */
    public void onDisconnect(String connectionId) {
        TrackedSession s = sessions.remove(connectionId);
        if (s != null) {
            s.state = SessionState.DISCONNECTED;
            s.playbackActive = false;
            s.sessionId = null;
            log.debug("[SessionTracker] Disconnected {}", connectionId);
        }
    }

    /**
     * Get the current active session with known sessionId.
     *
     * @return the most recently active session, or null if none
     */
    public ActiveSessionInfo getActiveSession() {
        TrackedSession best = null;
        for (TrackedSession s : sessions.values()) {
            if (s.state == SessionState.DISCONNECTED || s.state == SessionState.STALE) continue;
            if (s.sessionId == null) continue;
            if (best == null || s.lastActivityAt > best.lastActivityAt) {
                best = s;
            }
        }
        if (best == null) return null;
        return new ActiveSessionInfo(best.connectionId, best.sessionId, best.state.name().toLowerCase());
    }

    /**
     * Get the unavailable reason if no active session is available.
     */
    public String getUnavailableReason() {
        if (sessions.isEmpty()) return "no_active_session";
        for (TrackedSession s : sessions.values()) {
            if (s.state != SessionState.DISCONNECTED) {
                return "session_id_unknown";
            }
        }
        return "no_active_session";
    }

    /** Get count of tracked connections. */
    public int size() {
        return sessions.size();
    }

    private void reapStale() {
        long now = System.currentTimeMillis();
        for (Map.Entry<String, TrackedSession> entry : sessions.entrySet()) {
            TrackedSession s = entry.getValue();
            if (s.state == SessionState.DISCONNECTED) {
                sessions.remove(entry.getKey());
                continue;
            }
            if (now - s.lastActivityAt > staleTimeoutMs) {
                if (s.state != SessionState.STALE) {
                    log.info("[SessionTracker] Connection {} went stale ({}s idle)",
                            s.connectionId, (now - s.lastActivityAt) / 1000);
                    s.state = SessionState.STALE;
                }
            }
        }
    }

    // ── Inner types ─────────────────────────────────────────────

    enum SessionState {
        CONNECTED, PLAYING, STALE, DISCONNECTED
    }

    static class TrackedSession {
        final String connectionId;
        final String channel;
        volatile String sessionId;
        volatile boolean playbackActive;
        volatile long lastActivityAt;
        volatile SessionState state;

        TrackedSession(String connectionId, String channel) {
            this.connectionId = connectionId;
            this.channel = channel;
            this.lastActivityAt = System.currentTimeMillis();
            this.state = SessionState.CONNECTED;
        }
    }

    /** Immutable result for getActiveSession(). */
    public static class ActiveSessionInfo {
        public final String connectionId;
        public final String sessionId;
        public final String state;

        ActiveSessionInfo(String connectionId, String sessionId, String state) {
            this.connectionId = connectionId;
            this.sessionId = sessionId;
            this.state = state;
        }
    }
}
